feat(ButtonIcon): refactor component

marcoskolodny opened the pull request:
Issue: [Link](https://jira.tid.es/browse/WEB-1721)

In this PR, we create the new version of IconButton component. After this is merged, we will start adapting all the Mistica components in order to use the updated button in them

marcoskolodny commented on src/__screenshot_tests__/__image_snapshots__/icon-button-screenshot-test-tsx-icon-button-small-1-snap.png:
Notice the space around the circle. This is intended: the minimum touchable area is 48px

aweell replied:
In a conversation with @yceballost we were discussing that for pointer devices the 48px minimum area is not working fine in small sizes (causing a disconnection between the area where the pointer appears vs the hover state appears). Can this minimum area apply only for touch devices?

In pointer devices the minimum target area to pass AA is 24x24

https://www.w3.org/WAI/WCAG22/Understanding/target-size-minimum.html

marcoskolodny replied:
Updated

marcoskolodny commented on src/__screenshot_tests__/icon-button-screenshot-test.tsx:
I'm not testing skins/theme variant combinations here because those scenarios are covered in the private skins story

marcoskolodny commented on src/__stories__/icon-button-story.tsx:
```diff
@@ -1,31 +1,81 @@
 };
 
-const handleOnPress = () => window.alert('pressed!');
+const handleOnPress = () => console.log('pressed!');
```
This allows you to see the pressed interaction in the button in a more friendly way (without an alert that pops up).

marcoskolodny commented on src/icon-button.tsx:
```diff
@@ -144,26 +154,200 @@ const RawIconButton: React.FC<Props> = (props) => {
 // Used internally by Mistica's components to avoid styles collisions
-export const BaseIconButton = (props: Props): JSX.Element => {
+
+export const BaseIconButton = (props: IconButtonProps): JSX.Element => {
```
This internal component is only used in some input fields. I believe that with this new version of the IconButton, it may not be really necessary anymore.

Once we update all the Mistica components with the new IconButton, I will remove it if it's possible.

marcoskolodny commented on src/__screenshot_tests__/__image_snapshots__/private-skin-components-screenshot-test-tsx-components-in-blau-1-snap.png:
@aweell you can use these screenshots to validate if the tokens are correct in every possible scenario (different skins, button type/variant, theme variant, dark mode).

I've also added here some examples forcing the overMedia prop in a black background (despite the fact it's an internal prop, it's good to test it here because this is a private story).

aweell commented on src/icon-button.css.ts:
```diff
@@ -1,7 +1,425 @@
+    'danger-default-inverse': {
+        vars: {
+            [background]: skinVars.colors.buttonLinkDangerBackgroundInverse,
+            [backgroundHover]: skinVars.colors.backgroundContainerBrandHover,
+            [backgroundActive]: skinVars.colors.backgroundContainerBrandPressed,
```
```suggestion
            [backgroundHover]: skinVars.colors.backgroundContainerHover,
            [backgroundActive]: skinVars.colors.backgroundContainerPressed,
```

aweell commented on src/icon-button.css.ts:
```diff
@@ -1,7 +1,425 @@
+    'danger-soft-inverse': {
+        vars: {
+            [background]: skinVars.colors.buttonLinkDangerBackgroundInverse,
+            [backgroundHover]: skinVars.colors.backgroundContainerBrandHover,
+            [backgroundActive]: skinVars.colors.backgroundContainerBrandPressed,
```
```suggestion
            [backgroundHover]: skinVars.colors.backgroundContainerHover,
            [backgroundActive]: skinVars.colors.backgroundContainerPressed,
```

marcoskolodny commented on src/__screenshot_tests__/__image_snapshots__/icon-button-screenshot-test-tsx-icon-button-small-in-desktop-1-snap.png:
@aweell I've fixed the touchable area and added this screenshot to verify that in a desktop device, the minimum touchable area is not 48px (only in touchable devices, as you can see in [this screenshot](https://github.com/Telefonica/mistica-web/pull/1022/files#diff-7a8db3f3a5567b99056660ed0c855bb146d349e133cbba40e6b4568ba3010f26))

pladaria commented on src/icon-button.css.ts:
```diff
@@ -1,7 +1,511 @@
+export const isLoading = style({});
+export const overlayContainer = style({});
+
+const transitionTiming = '0.1s ease-in-out';
```
I think this is too fast, the transition is barely noticeable. Consider something like 0.3s

marcoskolodny replied:
this was an old bug that we never noticed until now. Good catch! It should be fixed now

marcoskolodny commented on src/counter.css.ts:
```diff
@@ -3,7 +3,10 @@ import * as mq from './media-queries.css';
-const transitionTiming = '0.2s ease-in-out';
+const colorTransitionTiming = '0.1s ease-in-out';
+const scaleTransitionTiming = '0.2s ease-in-out';
+const contentTransitionTiming = '0.3s cubic-bezier(0.77, 0, 0.175, 1)';
```
unrelated: animations were not consistent with specs in counter component

marcoskolodny commented on src/button.css.ts:
```diff
@@ -5,7 +5,8 @@ import * as mq from './media-queries.css';
 
-const transitionTiming = '0.1s ease-in-out';
+const colorTransitionTiming = '0.1s ease-in-out';
+const contentTransitionTiming = '0.3s cubic-bezier(0.77, 0, 0.175, 1)';
```
unrelated: animations were not consistent with specs in buttons

aweell commented:
The bleed properties seem to return a positive value instead of negative?

https://tinyurl.com/2yj7ercx

marcoskolodny commented on playroom/snippets.tsx:
```diff
@@ -581,7 +591,7 @@ const popoverSnippets = [
         code: `
         <Popover
-          target="some target, this can be any component"
+          target={<Text3>some target, this can be any component</Text3>}
```
unrelated: code cleanup

yceballost commented on src/icon-button.tsx:
Two problems found:
1. Badge is positioned below the iconButton. Should be positioned over the circle
2. When the variant is "undefined" (do not match with the real ones) a blue color appears (I think is the default blue color for links in the browser). Only appears with href

[Playroom example](https://mistica-izqwwj0sx-tuentisre.vercel.app/playroom#?code=N4Igxg9gJgpiBcIA8AFATjAbgSxgdwBUIIAbAZwAIB6APgB0A7JAIQgA8KAHAQyimwYBzALzAAjADYAvvQYUKSAJIMSAmBTI8wMUQBYADDMbz5SAMoAXbmADWGrTuAGjck6ea9B6zNxIBXRwAmFzc3JUgGZj8LCwgGY1C3ACNrG0E0CD8GKABhUgg0UUgSArIAOiS0bmypBMT5BnwCbiS6%2BrIAW18SNsSLAE9OHToQSuqoEd7QnzRsaothEbIIADMLSdd6igALDBXFkDwjssFiQRIYMpgyDa35RQjRB7iUEm5%2BgDFsEguoWs3QrQpkgqB4oF5ZIkWJ5vL4AqJgpD6uE4lEYnEpm5GoQWpiTJ1unj5AMhgdGtEqj0QESKDM5gwFiMsrAVmoJtSAYldjB9iMjngTmcLlcbhy7hRngwnhFXu8vj8YH88UDOQpQTCkWEwV5aXCgvpDJrQijItFYvFVW4CT8aS8MGQbpaTCThiByRZKbc7nT5gdlmsvVtJdKXm9Pt9fv8tiqoerwTAje4Ybr-EEQsjJWjzTTrT0ncTBq73Z6xd7uLNfUtSNh2TTg8BJbLwwqlfmKHbro7o4m1dqE8C%2Byn4cBEXiTVmMW3czSXWSYBTfIH6j6GQcWdw-CR1qWg48GzKw-LI7aGOhO8qeyC%2B0aQZZUje77Z7NZHM5L9ex5mzZPxSlbOlMmyPISkKYBilKCoqhqGlsWaVopy6G021nEYxmyJdEhXRkQH9bcaW5XlDmOU4IHOS5rgw0J60bQ8I0VKN6hjY04whAcNU-CIJwtcVYNxBDCWQws5wXKkaSwg5mR5NlKLcAiDn5QVSOFCid3qaiDzlOjW27YEWP7VVoXjHtTC-dFuLuac2w7B0ZyEkZi0XVTMPLelsNwmSTHU0NNJbBjEiYsI9PfdinXHb9zK2SzxRQt15w9RyxJcyscOrWs2y809aN8k8zxsp0ApMK8QqhD9QtM7N%2BKQ6K7NikSPPkcSRnXTc8PSvcaJ848rNPe0u0Yy8graW8rFsJEQWUVRGlkK92FkEAABoQAsbYYA6a4EAAbRAABZCAcDIKw0BAABdRa8BrZayE2gBmCR9GOqQgA)

marcoskolodny replied:
1. I found out that this problem is caused by Badge component, not the IconButton ([Playroom example](https://mistica-web.vercel.app/playroom#?code=N4Igxg9gJgpiBcIA8AhCAPABABwIZSgEsA7AcwF5gBmAJgF8A%2BAHWM01X1JkwDdcAbAK4xK9ZqzbsiPTAGcALgE9%2BI4MBwRZheYQjF4mJiABOMfrh08YRzHUYtJkpNIePJC5auCu3kgO6EUPIAFgYAjAAMEQA0Pr6YwTCEpMHy4VGxEvGYAEa4YADWpMYQgsRQBkY5QtYgmdl2cWwA9OKOSM3SbezNKJww4h1o6OJ1ICEwALYwsggA2iAAaoQ8EAD6AHIwfiAAutEgAUHBs-BztBG7dEA)). I've created a task to fix it: https://jira.tid.es/browse/WEB-1738
2. This is not a problem. `"undefined"` is not a valid value for the variant prop, and TS would complain about it. I would recommend you to check in a code editor that the playroom examples are actually valid usages of the components.

yceballost commented on src/icon-button.tsx:
```diff
@@ -1,9 +1,63 @@
+
+type OldProps = BaseOldProps & ExclusifyUnion<HrefProps | ToProps | OnPressProps | MaybeProps>;
 
 const ICON_SIZE_1 = 24;
```
what is "1" in this name? 🤔

marcoskolodny replied:
This is old code. I left everything related to old implementation untouched

yceballost commented on src/icon-button.css.ts:
```diff
@@ -1,7 +1,513 @@
+            [background]: skinVars.colors.inverse,
+            [backgroundHover]: skinVars.colors.backgroundContainerHover,
+            [backgroundActive]: skinVars.colors.backgroundContainerPressed,
+            [iconColor]: 'black',
```
why black? 🤔

aweell replied:
This is the same approach we're using in the current top actions definition, since inverse backgrounds are always white no matter the color-scheme, there's no available token that stays dark in both scenarios and is semantically suitable for this use case.

aweell commented on src/icon-button.css.ts:
```diff
@@ -1,7 +1,513 @@
+            [background]: 'transparent',
+            [backgroundHover]: skinVars.colors.backgroundContainerHover,
+            [backgroundActive]: skinVars.colors.backgroundContainerPressed,
+            [iconColor]: skinVars.colors.textButtonSecondary,
```
```suggestion
            [iconColor]: skinVars.colors.textLink,
```

aweell commented on src/icon-button.css.ts:
```diff
@@ -1,7 +1,513 @@
+            [background]: 'transparent',
+            [backgroundHover]: skinVars.colors.backgroundContainerBrandHover,
+            [backgroundActive]: skinVars.colors.backgroundContainerBrandPressed,
+            [iconColor]: skinVars.colors.textButtonSecondaryInverse,
```
```suggestion
            [iconColor]: skinVars.colors.textLinkInverse,
```

aweell commented on src/icon-button.css.ts:
```diff
@@ -1,7 +1,513 @@
+            [background]: 'transparent',
+            [backgroundHover]: skinVars.colors.backgroundContainerHover,
+            [backgroundActive]: skinVars.colors.backgroundContainerPressed,
+            [iconColor]: skinVars.colors.textButtonSecondary,
```
```suggestion
            [iconColor]: skinVars.colors.textLink,
```

aweell commented on src/icon-button.css.ts:
```diff
@@ -1,7 +1,513 @@
+            [background]: skinVars.colors.brandLow,
+            [backgroundHover]: skinVars.colors.backgroundContainerHover,
+            [backgroundActive]: skinVars.colors.backgroundContainerPressed,
+            [iconColor]: skinVars.colors.textButtonSecondary,
```
```suggestion
            [iconColor]: skinVars.colors.textLink,
```

aweell commented on src/icon-button.css.ts:
```diff
@@ -1,7 +1,513 @@
+            [background]: skinVars.colors.backgroundContainerBrandOverInverse,
+            [backgroundHover]: skinVars.colors.backgroundContainerBrandHover,
+            [backgroundActive]: skinVars.colors.backgroundContainerBrandPressed,
+            [iconColor]: skinVars.colors.textButtonSecondaryInverse,
```
```suggestion
            [iconColor]: skinVars.colors.textLinkInverse,
```

aweell commented on src/icon-button.css.ts:
```diff
@@ -1,7 +1,513 @@
+            [background]: skinVars.colors.brandLow,
+            [backgroundHover]: skinVars.colors.backgroundContainerHover,
+            [backgroundActive]: skinVars.colors.backgroundContainerPressed,
+            [iconColor]: skinVars.colors.textButtonSecondary,
```
```suggestion
            [iconColor]: skinVars.colors.textLink,
```

aweell reviewed (dismissed):
In order to maintain brand color in dark mode for soft and default variants of type = brand we have updated the specs with the following changes.

marcoskolodny commented:
> In order to maintain brand color in dark mode for soft and default variants of type = brand we have updated the specs with the following changes.

@aweell I've updated the tokens according to your comments :)

marcoskolodny commented on src/icon-button.tsx:
```diff
@@ -1,9 +1,63 @@
+    trackingEvent?: TrackingEvent | ReadonlyArray<TrackingEvent>;
+    /** "data-" prefix is automatically added. For example, use "testid" instead of "data-testid" */
+    dataAttributes?: DataAttributes;
+    newTab?: boolean;
```
Yes, I know. But moving that now is a breaking change, and given that we are gonna remove this code in the next major, I think we can wait until then

marcoskolodny commented on src/icon-button.css.ts:
```diff
@@ -1,7 +1,513 @@
+const backgroundActive = createVar();
+const iconColor = createVar();
+
+export const vars = {iconColor, background, backgroundHover, backgroundActive};
```
good catch. It was old code

marcoskolodny commented on src/__stories__/icon-button-story.tsx:
```diff
@@ -1,31 +1,81 @@
+            options: ['default', 'soft', 'solid'],
+            control: {type: 'select'},
+        },
+        theme: {
```
updated

marcoskolodny commented on src/icon-button.tsx:
```diff
@@ -144,26 +154,200 @@ const RawIconButton: React.FC<Props> = (props) => {
+    TouchableElement,
+    IconButtonProps & {isOverMedia?: boolean}
+>((props, ref) => {
+    if (props.Icon) {
```
done

marcoskolodny commented on src/icon-button.css.ts:
```diff
@@ -1,7 +1,513 @@
+        display: 'inline-flex',
+    }),
+    {
+        zIndex: 1,
```
Old code. Removed

marcoskolodny commented on src/icon-button.tsx:
```diff
@@ -144,26 +154,193 @@ const RawIconButton: React.FC<Props> = (props) => {
+                    <Icon size={styles.iconSize[buttonSize]} color="currentColor" />
+                </div>
+
+                <div
```
Yes. The behavior is equivalent to the Button components

marcoskolodny commented on src/icon-button.css.ts:
```diff
@@ -1,7 +1,513 @@
+};
+
+const pointerArea = {
+    default: `calc(max(${minButtonArea.pointer}, ${iconContainerSize.default}))`,
+    small: `calc(max(${minButtonArea.pointer}, ${iconContainerSize.small}))`,
```
Discussed offline. Added fallback to iconContainerSize if max() is not supported

marcoskolodny commented on src/icon-button.tsx:
```diff
@@ -144,26 +154,193 @@ const RawIconButton: React.FC<Props> = (props) => {
+    'aria-label'?: string;
+    small?: boolean;
+    type?: 'neutral' | 'brand' | 'danger';
+    variant?: 'default' | 'solid' | 'soft';
```
discussed offline. Changed to BackgroundType

tuentisre commented:
:tada: This PR is included in version 14.45.0 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/@telefonica/mistica/v/14.45.0)
- [GitHub release](https://github.com/Telefonica/mistica-web/releases/tag/v14.45.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
